ouroboros-network/src/Ouroboros/Network/ChainProducerState.hs

```diff
@@ -248,11 +256,14 @@ rollback p (ChainProducerState c rs) =
     ChainProducerState <$> Chain.rollback p c
                        <*> pure rs'
   where
-    rs' = [ if pointSlot p' > pointSlot p
+    rs' = [ if betterPoint p' p
               then r { readerPoint = p, readerNext = ReaderBackTo }
               else r
           | r@ReaderState { readerPoint = p' } <- rs ]
-
+    betterPoint Chain.Origin _ = False
+    betterPoint _ Chain.Origin = True
+    betterPoint (Chain.Point p') (Chain.Point p'') =
+      Chain.pointSlot p' > Chain.pointSlot p''
```

Comment:
Why do we need this? We have an `Ord` instance for `Point`.

Reply:
That `Ord` instance is derived, so it compares on the hash as well. For points at equal slots with different hashes, one would be determined to be better according to the hash ordering, which is weird and wrong. With this implementation `betterPoint`, the new point is discarded if its slot is not strictly greater.